### PR TITLE
Core/PacketIO: CalendarPackets:

### PR DESCRIPTION
**Pull request by Golrag**
CMSG_CALENDAR_GET
CMSG_CALENDAR_GET_EVENT
CMSG_CALENDAR_GUILD_FILTER
CMSG_CALENDAR_ADD_EVENT
CMSG_CALENDAR_UPDATE_EVENT
CMSG_CALENDAR_REMOVE_EVENT
CMSG_CALENDAR_COPY_EVENT // Needs test
CMSG_CALENDAR_GET_NUM_PENDING
CMSG_CALENDAR_EVENT_SIGN_UP
CMSG_CALENDAR_REMOVE_INVITE
CMSG_CALENDAR_EVENT_STATUS
CMSG_SET_SAVED_INSTANCE_EXTEND
CMSG_CALENDAR_EVENT_MODERATOR_STATUS

SMSG_CALENDAR_EVENT_INVITE
SMSG_CALENDAR_SEND_CALENDAR
SMSG_CALENDAR_SEND_EVENT
SMSG_CALENDAR_EVENT_INVITE_ALERT
SMSG_CALENDAR_EVENT_INVITE_STATUS
SMSG_CALENDAR_EVENT_INVITE_REMOVED
SMSG_CALENDAR_EVENT_INVITE_MODERATOR_STATUS
SMSG_CALENDAR_EVENT_INVITE_REMOVED_ALERT
SMSG_CALENDAR_CLEAR_PENDING_ACTION
SMSG_CALENDAR_EVENT_UPDATED_ALERT
SMSG_CALENDAR_EVENT_REMOVED_ALERT
SMSG_CALENDAR_SEND_NUM_PENDING
SMSG_CALENDAR_COMMAND_RESULT
SMSG_CALENDAR_RAID_LOCKOUT_UPDATED
SMSG_CALENDAR_EVENT_INITIAL_INVITES
SMSG_CALENDAR_EVENT_INVITE_STATUS_ALERT // NYI
SMSG_CALENDAR_EVENT_INVITE_NOTES_ALERT // NYI
SMSG_CALENDAR_EVENT_INVITE_NOTES // NYI

Any feedback & help is welcome ! :)

### CODE REVIEW

**Comment by Shauren** [comment]
So bad... you deleted ALL anti-exploit checks (such as invite limit to prevent using up entire memory on vector resize)

File: src/server/game/Calendar/CalendarMgr.cpp [hunk 1]
```diff
@@ -22,15 +22,16 @@
 #include "GuildMgr.h"
 #include "ObjectAccessor.h"
 #include "Opcodes.h"
+#include "CalendarPackets.h"
 
 CalendarInvite::~CalendarInvite()
 {
-    sCalendarMgr->FreeInviteId(_inviteId);
+    sCalendarMgr->FreeInviteId(_inviteID);
```
Comment (jackpoz) [inline src/server/game/Calendar/CalendarMgr.cpp]: are these d to D changes needed for this PR or are they codestyle changes ?

Comment (Golrag) [inline src/server/game/Calendar/CalendarMgr.cpp]: just codestyle changes, should I keep the old ones ?

Comment (Shauren) [inline src/server/game/Calendar/CalendarMgr.cpp]: yes

Comment (Golrag) [inline src/server/game/Calendar/CalendarMgr.cpp]: Okay :)
What about the names and DB changes ? Keep the old ones as well ?

Comment (DDuarte) [inline src/server/game/Calendar/CalendarMgr.cpp]: Those are good changes, keep them :+1:

File: src/server/game/Server/Packets/CalendarPackets.cpp [hunk 2]
```diff
@@ -0,0 +1,462 @@
...
+    _worldPacket << uint32(Invites.size());
+    for (auto const& invite : Invites)
+        _worldPacket << invite;
+
+    _worldPacket.FlushBits();
+
+    _worldPacket.WriteBits(EventName.size(), 8);
+    _worldPacket.WriteBits(Description.size(), 11);
+
+    _worldPacket.WriteString(EventName);
```
Comment (DDuarte) [inline src/server/game/Server/Packets/CalendarPackets.cpp]: missing a FlushBits() before this call

File: src/server/game/Handlers/CalendarHandler.cpp [hunk 3]
```diff
@@ -727,16 +547,15 @@ void WorldSession::SendCalendarRaidLockoutUpdated(InstanceSave const* save)
...
-    data << uint32(save->GetMapId());
-    data << uint32(save->GetDifficultyID());
-    data << uint32(0); // Amount of seconds that has changed to the reset time
-    data << uint32(save->GetResetTime() - currTime);
-    SendPacket(&data);
+    WorldPackets::Calendar::CalendarRaidLockoutUpdated packet;
+    packet.DifficultyID = save->GetDifficultyID();
+    packet.MapID = save->GetMapId();
+    packet.NewTimeRemaining = 0; // FIXME
+    packet.OldTimeRemaining = 0; // FIXME
```
Comment (DDuarte) [inline src/server/game/Handlers/CalendarHandler.cpp]: You can use `save->GetResetTime() - currTime` like we were doing.

File: src/server/game/Handlers/CalendarHandler.cpp [hunk 4]
```diff
@@ -563,61 +411,45 @@ void WorldSession::HandleCalendarEventRsvp(WorldPacket& recvData)
...
         {
-            invite->SetStatus((CalendarInviteStatus)status);
-            // not sure if we should set response time when moderator changes invite status
-            //invite->SetStatusTime(time(NULL));
+            invite->SetStatus((CalendarInviteStatus)calendarEventStatus.Status);
 
             sCalendarMgr->UpdateInvite(invite);
             sCalendarMgr->SendCalendarEventStatus(*calendarEvent, *invite);
-            sCalendarMgr->SendCalendarClearPendingAction(invitee);
+            //sCalendarMgr->SendCalendarClearPendingAction(calendarEventStatus.Guid);
```
Comment (DDuarte) [inline src/server/game/Handlers/CalendarHandler.cpp]: Reason?

File: src/server/game/Handlers/CalendarHandler.cpp [hunk 5]
```diff
@@ -526,35 +382,27 @@ void WorldSession::HandleCalendarEventSignup(WorldPacket& recvData)
...
         {
-            invite->SetStatus(CalendarInviteStatus(status));
-            invite->SetStatusTime(time(NULL));
+            invite->SetStatus(CalendarInviteStatus(calendarEventRSVP.Status));
+            invite->SetResponseTime(time(NULL));
 
             sCalendarMgr->UpdateInvite(invite);
             sCalendarMgr->SendCalendarEventStatus(*calendarEvent, *invite);
-            sCalendarMgr->SendCalendarClearPendingAction(guid);
+            //sCalendarMgr->SendCalendarClearPendingAction(guid);
```
Comment (DDuarte) [inline src/server/game/Handlers/CalendarHandler.cpp]: Reason?

File: src/server/game/Handlers/CalendarHandler.cpp [hunk 6]
```diff
@@ -47,196 +47,118 @@ Copied events should probably have a new owner
...
-        else
-        {
-            data << uint8(0);
-            data << (*itr)->GetSenderGUID().WriteAsPacked();
-        }
+        WorldPackets::Calendar::CalendarSendCalendarInviteInfo inviteInfo;
+        inviteInfo.EventID = invite->GetEventId();
+        inviteInfo.InviteID = invite->GetInviteId();
+        inviteInfo.InviterGuid = invite->GetSenderGUID();
+        inviteInfo.InviteType = 1; // Always 1 ?
```
Comment (DDuarte) [inline src/server/game/Handlers/CalendarHandler.cpp]: Looking at the old code only, this seems to be 1 it's a guild event and 0 when it's not.

File: src/server/game/Server/Packets/CalendarPackets.cpp [hunk 7]
```diff
@@ -0,0 +1,462 @@
...
+    data << uint64(eventInfo.EventID);
+    data << uint8(eventInfo.EventType);
+    data.AppendPackedTime(eventInfo.Date);
+    data << uint32(eventInfo.Flags);
+    data << int32(eventInfo.TextureID);
+    data << eventInfo.EventGuildID;
+    data << eventInfo.OwnerGuid;
+
+    data.FlushBits();
+    data.WriteBits(eventInfo.EventName.size(), 8);
```
Comment (MitchesD) [inline src/server/game/Server/Packets/CalendarPackets.cpp]: FlushBits should be after WriteBits, swap it

File: src/server/game/Server/Packets/CalendarPackets.cpp [hunk 8]
```diff
@@ -0,0 +1,462 @@
...
+
+    data << uint8(inviteInfo.Level);
+    data << uint8(inviteInfo.Status);
+    data << uint8(inviteInfo.Moderator);
+    data << uint8(inviteInfo.InviteType);
+
+    data.AppendPackedTime(inviteInfo.ResponseTime);
+
+    data.FlushBits();
+    data.WriteBits(inviteInfo.Notes.size(), 8);
```
Comment (MitchesD) [inline src/server/game/Server/Packets/CalendarPackets.cpp]: again swap

File: src/server/game/Server/Packets/CalendarPackets.cpp [hunk 9]
```diff
@@ -0,0 +1,462 @@
...
+
+void WorldPackets::Calendar::CalendarEventInvite::Read()
+{
+    _worldPacket >> EventID;
+    _worldPacket >> ModeratorID;
+
+    uint16 nameLen = _worldPacket.ReadBits(9);
+    Creating = _worldPacket.ReadBit();
+    IsSignUp = _worldPacket.ReadBit();
+    _worldPacket.FlushBits();
```
Comment (MitchesD) [inline src/server/game/Server/Packets/CalendarPackets.cpp]: remove that line, we use flushbits only when writting

File: src/server/game/Calendar/CalendarMgr.cpp [hunk 10]
```diff
@@ -554,105 +545,99 @@ void CalendarMgr::SendCalendarEvent(ObjectGuid guid, CalendarEvent const& calend
...
-        data << uint8(calendarEvent.IsGuildEvent() && calendarEvent.GetGuildId() == inviteeGuildId);
-        data << uint64(calendarInvite->GetInviteId());
-        data.AppendPackedTime(calendarInvite->GetStatusTime());
-        data << calendarInvite->GetText();
+        WorldPackets::Calendar::CalendarEventInviteInfo inviteInfo;
+        inviteInfo.Guid = inviteeGuid;
+        inviteInfo.Level = inviteeLevel;
+        inviteInfo.Status = calendarInvite->GetStatus();
+        inviteInfo.Moderator = calendarInvite->GetRank();
+        inviteInfo.InviteType = calendarEvent.IsGuildEvent(); // Correct ?
```
Comment (DDuarte) [inline src/server/game/Calendar/CalendarMgr.cpp]: data << uint8(calendarEvent.IsGuildEvent() && calendarEvent.GetGuildId() == inviteeGuildId);

File: src/server/game/Server/Packets/CalendarPackets.h [hunk 11]
```diff
@@ -0,0 +1,576 @@
...
+            ObjectGuid InviteGuid;
+            uint64 EventID = 0;
+            uint32 Flags = 0;
+            bool ClearPending = false;
+        };
+
+        class CalendarEventInviteModeratorStatus final : public ServerPacket
+        {
+        public:
+            CalendarEventInviteModeratorStatus() : ServerPacket(SMSG_CALENDAR_EVENT_INVITE_MODERATOR_STATUS, 29) { }
```
Comment (MitchesD) [inline src/server/game/Server/Packets/CalendarPackets.h]: 16 + 8 + 1 + 1 = 26 (ObjectGuid can have different size, so might be the reason why in sniff its 29)

File: sql/updates/characters/yyyy_mm_dd_xx_characters.sql [hunk 12]
```diff
@@ -0,0 +1,20 @@
+ALTER TABLE `calendar_invites`
```
Comment (DDuarte) [inline sql/updates/characters/yyyy_mm_dd_xx_characters.sql]: These changes need to be in the character_base.sql file as well. Use mysqldump to update that file.

File: src/server/game/Handlers/CalendarHandler.cpp (hunk 4 above)
Comment (Golrag) [inline src/server/game/Handlers/CalendarHandler.cpp]: Not sure if SMSG_CALENDAR_CLEAR_PENDING_ACTION should be called now. But I haven't really looked at it, so I will remove this change :)

File: src/server/game/Server/Packets/CalendarPackets.cpp [hunk 13]
```diff
@@ -0,0 +1,462 @@
...
+    _worldPacket << uint32(Flags);
+    _worldPacket.AppendPackedTime(Date);
+    _worldPacket << uint32(LockDate);
+    _worldPacket << EventGuildID;
+
+    _worldPacket << uint32(Invites.size());
+    for (auto const& invite : Invites)
+        _worldPacket << invite;
+
+    _worldPacket.FlushBits();
```
Comment (MitchesD) [inline src/server/game/Server/Packets/CalendarPackets.cpp]: not needed

**Comment by DDuarte** [comment]
Looking better, can we merge this?

**Comment by Shauren** [comment]
No, packet spoofing exploit is still there.

File: src/server/game/Server/Packets/CalendarPackets.cpp [hunk 14]
```diff
@@ -0,0 +1,460 @@
...
+    uint8 titleLength = _worldPacket.ReadBits(8);
+    uint16 descriptionLength = _worldPacket.ReadBits(11);
+
+    _worldPacket >> EventInfo.EventType;
+    _worldPacket >> EventInfo.TextureID;
+    EventInfo.Time = _worldPacket.ReadPackedTime();
+    _worldPacket >> EventInfo.Flags;
+
+    EventInfo.Invites.resize(_worldPacket.read<int32>());
```
Comment (Shauren) [inline src/server/game/Server/Packets/CalendarPackets.cpp]: You must never blindly trust what players send when it comes to sizes like this.

File: src/server/game/Server/Packets/CalendarPackets.h [hunk 15]
```diff
@@ -0,0 +1,576 @@
...
+        {
+        public:
+            CalendarSendCalendar() : ServerPacket(SMSG_CALENDAR_SEND_CALENDAR, 338) { }
+
+            WorldPacket const* Write() override;
+
+            time_t RaidOrigin = time_t(0);
+            time_t ServerTime = time_t(0);
+            time_t ServerNow = time_t(0);
+            std::vector<CalendarSendCalendarHolidayInfo> Holidays;
```
Comment (Shauren) [inline src/server/game/Server/Packets/CalendarPackets.h]: Why is this here? This packet no longer contains holiday info (it was moved to SMSG_DB_REPLY as Holidays is now a db2 storage)

File: src/server/game/Server/Packets/CalendarPackets.cpp [hunk 16]
```diff
@@ -0,0 +1,463 @@
...
+void WorldPackets::Calendar::CalendarGuildFilter::Read()
+{
+    _worldPacket >> MinLevel;
+    _worldPacket >> MaxLevel;
+    _worldPacket >> MaxRankOrder;
+}
+
+void WorldPackets::Calendar::CalendarAddEvent::Read()
+{
+    _worldPacket.FlushBits();
```
Comment (MitchesD) [inline src/server/game/Server/Packets/CalendarPackets.cpp]: not needed, flushbits is only for writing

File: src/server/game/Server/Packets/CalendarPackets.cpp [hunk 17]
```diff
@@ -0,0 +1,462 @@
...
+    _worldPacket >> EventInfo.EventType;
+    _worldPacket >> EventInfo.TextureID;
+    EventInfo.Time = _worldPacket.ReadPackedTime();
+    _worldPacket >> EventInfo.Flags;
+    uint32 count = _worldPacket.read<uint32>();
+
+    EventInfo.Title = _worldPacket.ReadString(titleLength);
+    EventInfo.Description = _worldPacket.ReadString(descriptionLength);
+
+    for (uint32 i = 0; i < count && i < CALENDAR_MAX_INVITES; i++)
```
Comment (Golrag) [inline src/server/game/Server/Packets/CalendarPackets.cpp]: @Shauren something like this ? ^^'

File: src/server/game/Server/Packets/CalendarPackets.cpp (hunk 16 above)
Comment (Golrag) [inline src/server/game/Server/Packets/CalendarPackets.cpp]: Done ! :p

**Comment by MitchesD** [comment]
@Golrag: could you please update it to 6.2? I would like to see this one merged ;-)

**Comment by Golrag** [comment]
@MitchesD soon :p
I found a 'small' bug. When I open the calendar, the week starts on sunday while it should start on monday ^^

**Comment by Golrag** [comment]
Updated to https://github.com/TrinityCore/TrinityCore/commit/776e4411c0ef5934aeb59f7b916a0b215b9ae80f

**Comment by Shauren** [comment]
@Golrag this is controlled by client locale - and since blizzard removed enGB everyone using that is now on enUS locale where their week starts from sunday (GAAAAAAAAAY)

**Comment by Golrag** [comment]
@Shauren Oh, okay :p

**Comment by Aokromes** [comment]
[ 25%] Building CXX object src/server/game/CMakeFiles/game.dir/Calendar/CalendarMgr.cpp.o
In file included from /home/travis/build/TrinityCore/TrinityCore/src/server/game/Calendar/CalendarMgr.cpp:18:
/home/travis/build/TrinityCore/TrinityCore/src/server/game/Calendar/CalendarMgr.h:211:56: fatal error: 
      field '_description' will be initialized after field '_lockDate'
      [-Wreorder]
            _date(date), _flags(flags), _title(title), _description(desc...
                                                       ^

**Comment by DDuarte** [comment]
Besides the merge problem that can be quickly solved, is there anything stopping this PR from being merged?

**Comment by Golrag** [comment]
Update to: https://github.com/TrinityCore/TrinityCore/commit/7f845294874a9cccfa0f13a6c7dbc9e88f5d3091
No bug reports/crashses are post in this topic, and I haven't found any so I guess it's good to go :)

File: src/server/game/Server/Packets/CalendarPackets.cpp [hunk 18]
```diff
@@ -0,0 +1,462 @@
...
+    _worldPacket >> EventInfo.EventID;
+    _worldPacket >> EventInfo.ModeratorID;
+    _worldPacket >> EventInfo.EventType;
+    _worldPacket >> EventInfo.TextureID;
+    EventInfo.Time = _worldPacket.ReadPackedTime();
+    _worldPacket >> EventInfo.Flags;
+
+    uint8 titleLen = _worldPacket.ReadBits(8);
+    uint16 descLen = _worldPacket.ReadBits(11);
+    _worldPacket.FlushBits();
```
Comment (MitchesD) [inline src/server/game/Server/Packets/CalendarPackets.cpp]: flush bits isnt needed when reading

**Comment by Golrag** [comment]
@DDuarte I forgot this change https://github.com/TrinityCore/TrinityCore/pull/14924/files#diff-e354d9cc799e748471cb44fc58cd2030R17 D:

**Pull request merged**